### PR TITLE
match: check scrutinee for partial application

### LINKED ISSUE #12093: warning [ignored-partial-application] does not fire on ignoring matches
## Context

Yesterday I had a frustrating debugging session with Olivier Martinot. We have a parser in our codebase (very simplified below), and some of the tests of our testsuite check that certain inputs are *rejected* by the parser as they should be.

```ocaml
  let parse input =
    if input = "correct" then `Correct
    else failwith "Parse error"
  
  let test_incorrect input =
    match parse input with
    | exception Failure _ -> true
    | _ -> false
```

Some changes we made to the machinery to display better error messages on parse errors had the surprising effect that our `test_incorrect` tests started failing (the inputs were passing instead of being rejected by the parser as expected). After a frustrating debug session, we realized that the cause was a change to our `parse` function to add a filename parameter:

```ocaml
  let parse ~filename input =
    ignore filename; (* feed it to the lexbuf etc. *)
    if input = "correct" then `Correct
    else failwith "Parse error"
  
  let test_incorrect input =
    match parse input with
    | exception Failure _ -> true
    | _ -> false
```

The bug is caused by the fact that `parse input` is now a partial application expecting a filename parameter, and in particular does not run any computation and never raises the `Failure` exception.

If we had written one of the following equivalent versions, we would have immediately caught our mistake thanks to Warning 5;

```ocaml
  let test_incorrect input =
    try let _ = parse input in true
             (* ^^^^^^^^^^^ *)
    with Failure _ -> false

  let test_incorrect input =
    try parse input; true
     (* ^^^^^^^^^^^ *)
    with Failure _ -> false

  (*
  Warning 5 [ignored-partial-application]: this function application is partial,
  maybe some arguments are missing.
  *)
```

(Note: other ways to avoid the issue would have been to add a typing annotation on the result of `parse input`, or to use a `result` type instead of an exception.)

## Proposal

I propose to extend warning 5 [ignored-partial-application] to matches that ignore their scrutinees, in addition to `ignore` and let-bindings that ignore their scrutinee.

### PR DESCRIPTION
**Pull request by nojb**
This PR proposes to check the scrutinee of a `match` expression in the same way that we do with the argument of `ignore` and the right hand side of `let _ = ...`. The check is only carried out if the only non-exceptional case of the pattern matching is a wildcard `_`. As with the existing cases, the warning can be disabled by annotating the scrutinee with a type constraint.

Fixes #12093

### CODE REVIEW

**Review by gasche — COMMENTED** [review]
Thanks for looking at this!

I first read the patch too quick and I thought that it would also warn on a `| _ -> ...` clause that had non-`_` clauses before in the same clause set. That would be wrong. This is not what your code does thanks to a well-placed `List.for_all`, but could the testsuite still include such a test?

The code looks correct, but now we have two sets of heuristics involved:
- Which scrutinee expression shapes trigger a warning (in particular, the fact that annotations silence it); this is in the `check_partial_application` logic that is shared with all other places that may raise this warning. Good!
- Which patterns are considered as ignoring their scrutinee. This is new code that is not shared with previous warning sites.

I would prefer if we could share a bit more code. In particular, doesn't the `let p = e` use-case also has logic to detect whether the pattern `p` counts as an ignore? Could we not share the code between `let` and `match`?

**Comment by nojb** [comment]
> Could we not share the code between `let` and `match`?

I pushed a commit with such an attempt. Let me know what you think.

File: typing/typecore.ml [hunk 1]
```diff
@@ -2655,6 +2655,18 @@ let check_partial_application ~statement exp =
...
+let pattern_needs_partial_application_check p =
+  let rec check : type a. a general_pattern -> bool = fun p ->
+    match p.pat_desc with
+    | Tpat_any ->
+        List.for_all (function (Tpat_constraint _, _, _) -> false | _ -> true) p.pat_extra
+    | Tpat_exception _ -> true
+    | Tpat_or (p1, p2, _) -> check p1 && check p2
+    | Tpat_value p -> check (p :> value general_pattern)
+    | _ -> false
+  in
```
Comment (gasche) [inline typing/typecore.ml:2668]: I think that it is slightly wrong that only `Tpat_any` checks the p.pat_extra. For example I would expect `((_ | _) : foo -> bar)` to disable the warning. I would expect the code to look at both pat_extra and pat_desc independently and perform a conjunction of the results.

Comment (nojb) [reply typing/typecore.ml:2668]: Yes, it is somewhat arbitrary; it was the existing logic in the `let _` case. I pushed a commit with your suggestion.

**Review by gasche — APPROVED** [review]
Approved. Thanks @nojb!

I think this needs a Changes entry, because it may create new warnings in existing codebases.

**Comment by stedolan** [comment]
This looks useful but I don't understand why it's specific to patterns that ignore their argument. Surely it is always a mistake worth warning about when the scrutinee of a match is a partial application? In cases that destructure the scrutinee, we'll get a type error as well, but mentioning that the scrutinee is partial would clarify things a lot. For instance:
```ocaml
# match List.map (fun x -> x + 1) with [x] -> 42 | _ -> 10;;
                                       ^^^
Error: This pattern should not be a list literal, the expected type is
       int list -> int list
```
The error message is in the wrong place, complaining about the pattern rather than the scrutinee. This is hard to improve (there's no real way for the compiler to guess which one is wrong), but if it also mentioned that the scrutinee was partial it would be easier to see the problem.

**Comment by nojb** [comment]
> Surely it is always a mistake worth warning about when the scrutinee of a match is a partial application?

Not always, eg this idiom in the compiler:

https://github.com/ocaml/ocaml/blob/bcffd84a9b0e38a3ca02a1ffd2de54541f3bcb9f/typing/env.ml#L2594-L2597

**Comment by gasche** [comment]
This needs a Changes entry before we can merge.

**Comment by nojb** [comment]
> This needs a Changes entry before we can merge.

This is now done (the PR should be squashed-merged).

**Comment by gasche** [comment]
Merged, thanks!

(cc the people who look at all-opam reports, @kit-ty-kate and @Octachron; this may break the build of warn-error projects)

**Pull request merged**